internal/workflow_shadower.go

```diff
@@ -0,0 +1,308 @@
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/facebookgo/clock"
+	"go.uber.org/cadence/.gen/go/cadence/workflowserviceclient"
+	"go.uber.org/cadence/.gen/go/shared"
+	"go.uber.org/cadence/internal/common"
+	"go.uber.org/cadence/internal/common/backoff"
+	"go.uber.org/cadence/internal/common/util"
+	"go.uber.org/zap"
+)
+
+var (
+	errInvalidTimeFilter = errors.New("should specify either TimeRange or Min, Max timestamp")
+)
+
+const (
+	statusInitialized int32 = 0
+	statusStarted     int32 = 1
+	statusStopped     int32 = 2
+)
+
+type (
+	// WorkflowShadowerOptions configs WorkflowShadower
+	WorkflowShadowerOptions struct {
+		Domain string
+
+		WorkflowQuery           string
+		WorkflowTypes           []string
+		WorkflowStatus          []string
+		WorkflowStartTimeFilter *TimeFilter
+		SamplingRate            float64
+
+		ExitCondition *WorkflowShadowerExitCondition
+
+		Logger *zap.Logger
+	}
+
+	// TimeFilter represents a time range filter
+	TimeFilter struct {
+		MinTimestamp time.Time
+		MaxTimestamp time.Time
+	}
+
+	// WorkflowShadowerExitCondition specifies when Shadower should stop shadowing and exit
+	WorkflowShadowerExitCondition struct {
+		ExpirationTime    time.Duration
+		MaxShadowingCount int
+	}
+
+	// WorkflowShadower retrieves and replays workflow history from Cadence service to determine if there's any nondeterministic changes in the workflow definition
+	WorkflowShadower struct {
+		service  workflowserviceclient.Interface
+		options  *WorkflowShadowerOptions
+		replayer *WorkflowReplayer
+
+		status     int32
+		shutdownCh chan struct{}
+		shutdownWG sync.WaitGroup
+
+		clock clock.Clock
+	}
+)
+
+// NewWorkflowShadower creates an instance of the WorkflowShadower
+func NewWorkflowShadower(
+	service workflowserviceclient.Interface,
+	options *WorkflowShadowerOptions,
+) (*WorkflowShadower, error) {
+	if err := options.validateAndPopulateFields(); err != nil {
+		return nil, err
+	}
+	return &WorkflowShadower{
+		service:  service,
+		options:  options,
+		replayer: NewWorkflowReplayer(),
+
+		status:     statusInitialized,
+		shutdownCh: make(chan struct{}),
+
+		clock: clock.New(),
+	}, nil
+}
+
+// RegisterWorkflow registers workflow function to replay
+func (s *WorkflowShadower) RegisterWorkflow(w interface{}) {
+	s.replayer.RegisterWorkflow(w)
+}
+
+// RegisterWorkflowWithOptions registers workflow function with custom workflow name to replay
+func (s *WorkflowShadower) RegisterWorkflowWithOptions(w interface{}, options RegisterWorkflowOptions) {
+	s.replayer.RegisterWorkflowWithOptions(w, options)
+}
+
+// Run starts WorkflowShadower in a blocking fashion
+func (s *WorkflowShadower) Run() error {
+	if !atomic.CompareAndSwapInt32(&s.status, statusInitialized, statusStarted) {
+		return errors.New("Workflow shadower already started")
+	}
+
+	return s.shadowWorker()
+}
+
+// Stop stops WorkflowShadower and wait up to one miniute for all goroutines to finish before returning
+func (s *WorkflowShadower) Stop() {
```

Comment:
Do you plan to define this in the interface?

Reply:
Probably not at this time. I will keep the implementation here for now and if user needs that capability after we rollout the feature, I can easily add it to the interface.